lib/src/rules/join_return_with_assignment_when_possible.dart

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.join_return_with_assignment_when_possible;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'Join return statement with assignment when possible.';
+
+const _details = r'''
+
+**DO** join return statement with assignment when possible.
+
+**BAD:**
+```
+class A {
+  B _lazyInstance;
+  static B get instance {
+    _lazyInstance ??= new B(); // LINT
+    return _lazyInstance;
+  }
+}
+```
+
+**GOOD:**
+```
+class A {
+  B _lazyInstance;
+  static B get instance => _lazyInstance ??= new B();
+}
+```
+
+''';
+
+Element _getElementFromAssignmentStatement(Statement node) {
+  final visitor = new _AssignmentStatementVisitor();
+  node.accept(visitor);
+  return visitor.element;
+}
+
+Element _getElementFromReturnStatement(Statement node) {
+  final visitor = new _ReturnStatementVisitor();
+  node.accept(visitor);
+  return visitor.element;
+}
+
+class JoinReturnWithAssignmentWhenPossible extends LintRule {
+  _Visitor _visitor;
+  JoinReturnWithAssignmentWhenPossible()
+      : super(
+            name: 'join_return_with_assignment_when_possible',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _AssignmentStatementVisitor extends SimpleAstVisitor {
+  Element element;
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    element =
+        DartTypeUtilities.getCanonicalElementFromIdentifier(node.leftHandSide);
+  }
+
+  @override
+  visitExpressionStatement(ExpressionStatement statement) {
+    statement.expression.accept(this);
+  }
+
+  @override
+  visitParenthesizedExpression(ParenthesizedExpression node) {
+    node.unParenthesized.accept(this);
+  }
+
+  @override
+  visitPostfixExpression(PostfixExpression node) {
+    element = DartTypeUtilities.getCanonicalElementFromIdentifier(node.operand);
+  }
+
+  @override
+  visitPrefixExpression(PrefixExpression node) {
+    element = DartTypeUtilities.getCanonicalElementFromIdentifier(node.operand);
+  }
+}
+
+class _ReturnStatementVisitor extends SimpleAstVisitor {
+  Element element;
+
+  @override
+  visitReturnStatement(ReturnStatement node) {
+    element =
+        DartTypeUtilities.getCanonicalElementFromIdentifier(node.expression);
```

Comment:
Defining a visitor with a single visit method seems unnecessary to me. Given the cost of polymorphic dispatch, it would be more efficient to test the type in the calling function.